### ISSUE TITLE
Uncaught TypeError. Cannot read property 'offsetWidth' of null in PathFindingLinkFactory calculateMatrixDimensions

### ISSUE THREAD

**Issue by flieks** [projectstorm/react-diagrams#671]
Getting this error when doing a recalc of the engine with all nodes.

![image](https://user-images.githubusercontent.com/5032389/86598338-3f3db200-bf9d-11ea-8279-01b50c8eaaa5.png)

stacktrace:
PathFindingLinkFactory.tsx:206 Uncaught TypeError: Cannot read property 'offsetWidth' of null
    at PathFindingLinkFactory._this.calculateMatrixDimensions (PathFindingLinkFactory.tsx:206)
    at PathFindingLinkFactory.calculateCanvasMatrix (PathFindingLinkFactory.tsx:101)
    at PathFindingLinkFactory.getCanvasMatrix (PathFindingLinkFactory.tsx:96)
    at PathFindingLinkFactory.calculateRoutingMatrix (PathFindingLinkFactory.tsx:142)
    at PathFindingLinkFactory.tsx:58
    at lodash.js:2750

**Comment by FavorMylikes** [comment]
It only  happens when you want to call `calculateRoutingMatrix` before dom render

**Comment by FavorMylikes** [comment]
From my experience, there is two ways to implement it. One of them is `setInterval` wait 200ms, in most of times, it works, or, use some third lib to monitor the dom finish.
But I'll suggest do not call `reroute` or some other function that depend on `calculateRoutingMatrix` in `render` function. And add `width` and `height`  in state as args.

We can look deep inside. 
`const maxX = Math.max(sumProps(maxXElement, ['x', 'width']), canvas.offsetWidth);`
This line is trying getting the width from canvas. and as the same time, the width is come from other config that like browser page or other `Component` states.
So why don't you try to set it in your own `Component`.

Any way, if there's any better way, glad to hear it.

**Comment by pSnehanshu** [comment]
@FavorMylikes Yes, there is a better way. react-diagrams is full of meaningless edge cases, hence it is better to completely avoid this library if you don't want to burn out. I will recommend https://www.npmjs.com/package/react-flow-renderer as an alternative.

**Comment by FavorMylikes** [comment]
Oh, cool, I'll try to integrate [dagre](https://github.com/dagrejs/dagre)

**Comment by Omario-r** [comment]
This works for me:

`function Diagram() {
  const [engine, setEngine] = useState()

  useEffect(() => {
    const engine = createEngine();
    /*  operations with engine */
    setEngine(engine)
  }, [])

  return (
    <div>
      {engine && <CanvasWidget engine={engine} />}
    </div>
  );
}`
